### DISCUSSION TITLE
Are typing errors in fastapi/starlite templates normal ?

### DISCUSSION THREAD

**Discussion by StitiFatah** [piccolo-orm/piccolo#709] [Q&A] [answered]
Hi, 

I'm trying piccolo with the Starlite template and have some typing errors (from pyright/pylance in both vscodium and neovim), for example with the `/tasks` endpoint :

```
@get("/tasks", tags=["Task"])
async def tasks() -> t.List[Task]:
    tasks = await Task.select().order_by(Task.id, ascending=False)
    return tasks
```

I get `Cannot access member "id" for type "Type[Task]"  Member "id" is unknown` , Is that because id isn't manually specified on the Table ? Why does the auto completion works fine in ipython ?

I also get  `Expression of type "Any | list[Table] | Table | Unknown | str | list[Unknown]" cannot be assigned to return type "List[Task]"` is unknown` for tasks.   Is that intended and should we use typing.cast for the moment ? 

I also get errors when using `Band.objects(Band.all_related())` with models that have foreign keys. 

But maybe that's an editor problem on my side ? 

Thanks in advance.

**Comment by dantownsend** [comment, marked as answer]
> I get Cannot access member "id" for type "Type[Task]" Member "id" is unknown , Is that because id isn't manually specified on the Table ? Why does the auto completion works fine in ipython ?

If a primary key isn't manually specified then Piccolo automatically adds `id`.

To fix this type error, you can add this to your table class:

```python
from piccolo.columns.column_types import Serial

class MyTable(Table):
    id: Serial
    ...
```

> I also get Expression of type "Any | list[Table] | Table | Unknown | str | list[Unknown]" cannot be assigned to return type "List[Task]" is unknown` for tasks. Is that intended and should we use typing.cast for the moment ?

Piccolo's `select` query returns a list of dictionaries. So the response type is something like `t.List[t.Dict]`.

To get a list of `Task` objects you need to use `objects` instead.

```python
@get("/tasks", tags=["Task"])
async def tasks() -> t.List[Task]:
    tasks = await Task.objects().order_by(Task.id, ascending=False)
    return tasks
```

We're working on improving the type annotations for Piccolo - you might need to add something like this for now:

```python
@get("/tasks", tags=["Task"])
async def tasks() -> t.List[Task]:
    tasks: t.List[Task] = await Task.objects().order_by(Task.id, ascending=False)
    return tasks
```

I have quite a large PR underway at the moment which uses Python's generics to improve the type responses of queries, so should help.

**Reply by StitiFatah** [reply]
Thanks for the response, 

>Piccolo's select query returns a list of dictionaries. So the response type is something like t.List[t.Dict].
To get a list of Task objects you need to use objects instead.

Yes it makes sense, I tried that but got the same error, but shouldn't the template replace select with objects ? I've just copy pasted it. 


>We're working on improving the type annotations for Piccolo - you might need to add something like this for now:
```
@get("/tasks", tags=["Task"])
async def tasks() -> t.List[Task]:
    tasks: t.List[Task] = await Task.objects().order_by(Task.id, ascending=False)
    return tasks
```

Ok I'll use that for now. 


Thanks for this library by the way, I'm considering switching from Django/DRF to Starlite and I really like Piccolo so far.

**Reply by dantownsend** [reply]
> Thanks for this library by the way, I'm considering switching from Django/DRF to Starlite and I really like Piccolo so far.

No worries, glad you're enjoying it.
